File: 이승연/Programmers/Level 3/입국심사.java
public class 입국심사 {
    class Solution {
        public long solution(int n, int[] times) {
            Arrays.sort(times);
            long left = 1;
            long right = (long)n*times[times.length-1];
            long answer = right;

            while(left<=right) {
                long middle = (left+right)/2;
                if (check(middle, n, times)) {
                    right = middle-1;
                    answer = Math.min(middle, answer);
                    continue;
                }
                left = middle+1;
            }
            return answer;
        }

        public boolean check(long time, long n, int[] times) {
            long count = 0;

            for(int i=0; i<times.length; i++) {
                count += time/times[i];
            }

            if (count>=n) return true;
            return false;
        }
    }
}

File: 정주완/Programmers/Level 3/입국심사.py
# 20m 20s

def solution(n, times):
    answer = 0
    
    # 파라메트릭 서치로 풀이
    start = min(times)
    end = max(times) * n
    
    while start <= end:
        mid = (start + end) // 2
        
        passed_people = 0 # mid 시간동안 심사 할 수 있는 사람의 수
        for time in times:
            passed_people += mid // time
        
        if passed_people >= n:
            answer = mid
            end = mid - 1
        else:
            start = mid + 1     
    
    return answer